ntoskrnl/io/pnpmgr/pnpmap.c

```diff
@@ -29,48 +35,114 @@ static UNICODE_STRING ConfigurationDataU = RTL_CONSTANT_STRING(L"Configuration D
 {
     { L"SerialController", L"*PNP0501\0", 0 },
-    { L"KeyboardController", L"*PNP0303\0", 0 },
-#if defined(SARCH_PC98)
-    { L"PointerController", L"*nEC1F00\0", 0 },
-#else
-    { L"PointerController", L"*PNP0F13\0", 0 },
-#endif
+    //{ L"KeyboardController", L"*PNP0303\0", 0 },
+    //{ L"PointerController", L"*PNP0F13\0", 0 },
+    { L"KeyboardPeripheral", NULL, 0 },
+    { L"PointerPeripheral", NULL, 0 },
     { L"ParallelController", L"*PNP0400\0", 0 },
     { L"FloppyDiskPeripheral", L"*PNP0700\0", 0 },
     { NULL, NULL, 0 }
 };
 
+static
+PNP_MAPPER_DEVICE_ID KeyboardMap[] =
+{
+    { L"XT_83KEY", L"*PNP0300\0" },
+    { L"PCAT_86KEY", L"*PNP0301\0" },
+    { L"PCXT_84KEY", L"*PNP0302\0" },
+    { L"XT_84KEY", L"*PNP0302\0" },
+    { L"101-KEY", L"*PNP0303\0" },
+    { L"OLI_83KEY", L"*PNP0304\0" },
+    { L"ATT_301", L"*PNP0304\0" },
+    { L"OLI_102KEY", L"*PNP0305\0" },
+    { L"OLI_86KEY", L"*PNP0306\0" },
+    { L"OLI_A101_102KEY", L"*PNP0309\0" },
+    { L"ATT_302", L"*PNP030a\0" },
+    { L"PCAT_ENHANCED", L"*PNP030b\0" },
+    { L"PC98_106KEY", L"*nEC1300\0" },
+    { L"PC98_LaptopKEY", L"*nEC1300\0" },
+    { L"PC98_N106KEY", L"*PNP0303\0" },
+    { NULL, NULL }
+};
+
+static
+PNP_MAPPER_DEVICE_ID PointerMap[] =
+{
+    { L"PS2 MOUSE", L"*PNP0F0E\0" },
+    { L"SERIAL MOUSE", L"*PNP0F0C\0" },
+    { L"MICROSOFT PS2 MOUSE", L"*PNP0F03\0" },
+    { L"LOGITECH PS2 MOUSE", L"*PNP0F12\0" },
+    { L"MICROSOFT INPORT MOUSE", L"*PNP0F02\0" },
+    { L"MICROSOFT SERIAL MOUSE", L"*PNP0F01\0" },
+    { L"MICROSOFT BALLPOINT SERIAL MOUSE", L"*PNP0F09\0" },
+    { L"LOGITECH SERIAL MOUSE", L"*PNP0F08\0" },
+    { L"MICROSOFT BUS MOUSE", L"*PNP0F00\0" },
+    { L"NEC PC-9800 BUS MOUSE", L"*nEC1F00\0" },
+    { NULL, NULL }
+};
+
 /* FUNCTIONS *****************************************************************/
 
+static
+CODE_SEG("INIT")
+PWSTR
+IopMapPeripheralId(
+    _In_ PUNICODE_STRING Value,
+    _In_ PPNP_MAPPER_DEVICE_ID DeviceList)
+{
+    ULONG i;
+    UNICODE_STRING CmpId;
+
+    if (!Value)
+        return NULL;
+
+    for (i = 0; DeviceList[i].TypeName; i++)
+    {
+        RtlInitUnicodeString(&CmpId, DeviceList[i].TypeName);
+
+        if (RtlCompareUnicodeString(Value, &CmpId, FALSE) == 0)
+            break;
+    }
+
+    return DeviceList[i].PnPId;
+}
+
 static
 CODE_SEG("INIT")
 PWSTR
 IopMapDetectedDeviceId(
     _In_ PUNICODE_STRING DetectId,
+    _In_ PUNICODE_STRING Value,
     _Out_ PULONG DeviceIndex)
 {
     ULONG i;
     UNICODE_STRING CmpId;
+    PCWSTR CurId;
 
     if (!DetectId)
         return NULL;
 
     for (i = 0; PnPMap[i].DetectId; i++)
     {
-        RtlInitUnicodeString(&CmpId, PnPMap[i].DetectId);
+        CurId = PnPMap[i].DetectId;
+        RtlInitUnicodeString(&CmpId, CurId);
 
         if (RtlCompareUnicodeString(DetectId, &CmpId, FALSE) == 0)
         {
             *DeviceIndex = PnPMap[i].Counter++;
+
+            if (wcscmp(CurId, L"KeyboardPeripheral") == 0)
```

Comment:
Why not
1. compare to a UNICODE_STRING, since you already have one in `CmpId` anyway, or
2. maybe even avoid this string comparison entirely by making the `PnPMap` predictable so that this can be an index check, or
3. maybe the most elegant: just put the KeyboardMap/PointerMap pointer inside the `PnPMap` array as a struct member?

(Side note, Windows doesn't seem to need a loop here at all since they just use the `CONFIGURATION_TYPE` returned by `IoQueryDeviceDescription`? Not super familiar with this code, so not sure if that idea is even applicable to ROS.)

Reply:
IoQueryDeviceDescription only takes its parameters as *inputs*: the pointer to `CONFIGURATION_TYPE` given to it is an input. (It's done that way, instead of giving the direct value, in order to tell the function what it should enumerate and what it shouldn't.)